R/labelled.R

```diff
@@ -0,0 +1,208 @@
+#'   tibble::tibble(y = letters[1:3]), tibble::tibble(field_name = "y", field_label = "Y Label")
+#' )
+#'
+#' make_labelled(supertbl)
+#'
+#' \dontrun{
+#' redcap_uri <- Sys.getenv("REDCAP_URI")
+#' token <- Sys.getenv("REDCAP_TOKEN")
+#'
+#' supertbl <- read_redcap_tidy(redcap_uri, token)
+#' make_labelled(supertbl)
+#' }
+#' @export
+make_labelled <- function(supertbl, format_labels = NULL) {
+
+  check_installed("labelled", reason = "to use `make_labelled()`")
+
+  if (is.null(format_labels)) {
+    formatter <- identity
+  } else if (is.function(format_labels)) {
+    formatter <- format_labels
+  } else if (format_labels == "default") {
+    formatter <- format_labels_default
+  } else {
+    # TODO: add informative error with cli
+    stop()
+  }
+
+  assert_data_frame(supertbl)
+  check_req_labelled_fields(supertbl)
+  check_req_labelled_metadata_fields(supertbl)
+
+  # Derive labels ----
+  main_labs <- c(
+    redcap_form_name = "REDCap Instrument Name",
+    redcap_form_label = "REDCap Instrument Description",
+    redcap_data = "Data",
+    redcap_metadata = "Metadata",
+    redcap_events = "Events and Arms Associated with this Instrument",
+    structure = "Repeating or Nonrepeating?",
+    data_rows = "# of Rows in Data",
+    data_cols = "# of Columns in Data",
+    data_size = "Data size in Memory",
+    data_na_pct = "% of Data Missing"
+  )
+
+  metadata_labs <- c(
+    field_name = "Variable / Field Name",
+    field_label = "Field Label",
+    field_type = "Field Type",
+    section_header = "Section Header",
+    field_note = "Field Note",
+    text_validation_type_or_show_slider_number = "Text Validation Type OR Show Slider Number",
+    text_validation_min = "Text Validation Min",
+    text_validation_max = "Text Validation Max",
+    identifier = "Identifier?",
+    branching_logic = "Branching Logic (Show field only if...)",
+    required_field = "Required Field?",
+    custom_alignment = "Custom Alignment",
+    question_number = "Question Number (surveys only)",
+    matrix_group_name = "Matrix Group Name",
+    matrix_ranking = "Matrix Ranking?",
+    field_annotation = "Field Annotation"
+  )
+
+  ## Set some predefined labels for data fields that aren't in the metadata
+  data_labs <- c(
+    redcap_repeat_instance = "REDCap Repeat Instance",
+    redcap_event = "REDCap Event",
+    redcap_arm = "REDCap Arm",
+    form_status_complete = "REDCap Form Status"
+  )
+
+  event_labs <- c(
+    redcap_event = "Event Name",
+    redcap_arm = "Arm Name",
+    arm_name = "Arm Description"
+  )
+
+  # Apply labels ----
+
+  # Utility function for label setting
+  # Set labels of tibble from named vector but don't fail if labels vector has
+  # variables that aren't in the data
+  safe_set_variable_labels <- function(data, labs) {
+    labs_to_keep <- intersect(names(labs), colnames(data))
+    labelled::set_variable_labels(data, !!!labs[labs_to_keep])
```

Comment:
Unsure if it's required to add an `@importFrom` since you're calling out the package here, but to keep consistency I'd throw `@importFrom laballed set_variable_labels` into the roxygen docs/`NAMESPACE`.

Reply:
When I do this and run `check()` I get:

```
checking package dependencies ... ERROR
Namespace dependency missing from DESCRIPTION Imports/Depends entries: ‘labelled’
```

Unless I'm missing something. I don't think you can have things in the `NAMESPACE` that aren't in `Imports/Depends` and `@importFrom` adds to the `NAMESPACE`.

Reply:
Right, in this case you would need to have it in the `DESCRIPTION` file via `use_package(__)`. We're currently doing that with `cli` which isn't a user-side thing, so I figure it doesn't hurt to do it for `labelled` as well, but I'm not really that opinionated one way or another.

We may also not need to do this anyway since you're checking for `labelled` installation in the new function anyway.

Reply:
I think the pattern we were envisioning is that the user only needs to have `labelled` installed if they want to use `make_labelled()`. So we put it in `Suggests` and check that the user has it installed at the top of `make_labelled()`.

The problem is that to use `@importFrom` we'd need to bump it from `Suggests` to `Imports` which would mean that it gets installed when a user installs `REDCapTidieR`.

Reply:
Ah, ok makes sense to me 👍